### PR TITLE
Feature/realm categories

### PR DESCRIPTION
**Pull request by lukasgolson**
This PR adds the realm categories feature as described in #89, solidifies instanced and temporary realms, and adds a teleporter block for easier player movement. 

For an example of how this teleport block might work with tutorial realms, execute the following commands `/realm blocks teleporter 1 true true punchablock punchablockschematic` to receive a block that will teleport you to a temporary, instanced realm containing the punchablock schematic.

### CODE REVIEW

File: mods/mc_worldmanager/realm/realmCategory.lua [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+        if (realm:get_data("students") == nil) then
+            realm:set_data("students", {})
+        end
+
+        if (realm:get_data("owner") == nil) then
+            realm:set_data("owner", {})
+        end
+
+        if (realm:get_data("students")[player:get_player_name()] ~= nil) then
```
Comment (pauldpickell) [inline mods/mc_worldmanager/realm/realmCategory.lua:82]: How are teachers treated? Need to consider the case that teachers should be able to join any realm (e.g., needing to help a student or grade an assignment/project), so perhaps they need to be added as owners?

File: mods/mc_worldmanager/realm/realmPlayerManagement.lua [hunk 2]
```diff
@@ -1,8 +1,16 @@
...
 ---TeleportPlayer
 ---Teleports a player to this realm.
----@param player objectRef
+---@param player
 function Realm:TeleportPlayer(player)
+    local realmCategory = self:getCategory()
+
+    local joinable, reason = realmCategory.joinable(self, player)
+
+    if (not joinable and not minetest.check_player_privs(player, { teacher = true })) then
```
Comment (pauldpickell) [inline mods/mc_worldmanager/realm/realmPlayerManagement.lua:10]: Okay, I see this now. So if I understand correctly, the teacher has to be added as either a student or an owner of the realm?

File: mods/mc_worldmanager/teleporterNode.lua [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                else
+                    realmObject = Realm:New(realmName, { x = 80, y = 80, z = 80 })
+                    realmObject:CreateGround()
+                    realmObject:CreateBarriersFast()
+                end
+
+                meta:set_int("realm", realmObject.ID)
+                minetest.swap_node(pos, { name = "mc_worldmanager:teleporter", param2 = math.ceil(math.sin(realmObject.ID) * 255) })
+            else
+                minetest.chat_send_player(clicker:get_player_name(), "This teleporter is broken. Please contact an instructor for assistance.")
```
Comment (pauldpickell) [inline mods/mc_worldmanager/teleporterNode.lua:89]: Might also be useful to indicate that the student can use the reporting function in the student notebook to report this issue.

**Review by pauldpickell — CHANGES_REQUESTED** [review]
Minor suggestions and some questions.

File: mods/mc_worldmanager/realm/realmCategory.lua (hunk 1 above)
Comment (lukasgolson) [reply mods/mc_worldmanager/realm/realmCategory.lua:82]: Users with the teachers privilege are assumed to be excluded from this check to make implementation details easier for any add-on mod. 

E.g., add-on mods should respect that teacher privileges override the check without creating a dependency.

See the teleportPlayer function in the realm player management file for an example of the assumption I'm making. 

I'll see if I can move this implementation from calling methods into this class.

File: mods/mc_worldmanager/realm/realmPlayerManagement.lua (hunk 2 above)
Comment (lukasgolson) [reply mods/mc_worldmanager/realm/realmPlayerManagement.lua:10]: See my above response to previous comment for context.

While this is true, it's not necessary. Players with the teacher privilege are not checked with this system under the assumption that calling functions will handle these players separately.

File: mods/mc_worldmanager/teleporterNode.lua (hunk 3 above)
Comment (lukasgolson) [reply mods/mc_worldmanager/teleporterNode.lua:89]: Good idea, I'll look into this.

**Pull request merged**